[mlir][SCF] Modernize `coalesceLoops` method to handle `scf.for` loops with iter_args

MaheshRavishankar opened the pull request:
As part of this extension this change also does some general cleanup

1) Make all the methods take `RewriterBase` as arguments instead of
   creating their own builders that tend to crash when used within
   pattern rewrites
2) Split `coalesePerfectlyNestedLoops` into two separate methods, one
   for `scf.for` and other for `affine.for`. The templatization didnt
   seem to be buying much there.

Also general clean up of tests.

qedawkins commented on mlir/lib/Dialect/SCF/Utils/Utils.cpp:
```diff
@@ -493,62 +520,130 @@ static LoopParams normalizeLoop(OpBuilder &boundsBuilder,
+delinearizeInductionVariable(RewriterBase &rewriter, Location loc,
+                             Value linearizedIv, ArrayRef<Value> ubs) {
+  auto isIndexType = [](Value v) { return v.getType().isa<IndexType>(); };
+  if (linearizedIv.getType().isa<IndexType>()) {
```
```suggestion
  if (linearizedIv.getType().isIndex()) {
```

qedawkins commented on mlir/lib/Dialect/SCF/Utils/Utils.cpp:
```diff
@@ -557,57 +652,151 @@ LogicalResult mlir::coalesceLoops(MutableArrayRef<scf::ForOp> loops) {
 
+LogicalResult mlir::coalesceLoops(MutableArrayRef<scf::ForOp> loops) {
+  if (loops.empty()) {
+    return success();
```
This looks inconsistent with the variant that take the rewriter, which returns failure if there are no loops
```
  if (loops.size() < 2)
    return failure();
```
This should either return failure too or the other should return success given no loops.

MaheshRavishankar replied:
Good catch. Made both failure fo rnow.

qedawkins commented on mlir/test/Dialect/SCF/transform-op-coalesce.mlir:
```diff
@@ -96,3 +101,150 @@ module attributes {transform.with_named_sequence} {
+//  CHECK-DAG: #[[MAP:.+]] = affine_map<()[s0, s1, s2] -> ((-s0 + s1) ceildiv s2)>
+//  CHECK-DAG: #[[MAP1:.+]] = affine_map<()[s0, s1, s2, s3, s4, s5, s6, s7, s8] -> ((((-s0 + s1) ceildiv s2) * ((-s3 + s4) ceildiv s5)) * ((-s6 + s7) ceildiv s8))>
+//  CHECK-DAG: #[[MAP3:.+]] = affine_map<()[s0, s1, s2] -> (s0 * s1 + s2)>
+//      CHECK: func.func @tensor_loops(
```
nit: CHECK-LABEL

MaheshRavishankar replied:
I leave it as CHECK for now. CHECK-LABEL resets the context that can be pierced in wierd ways. Also I think `CHECK-SAME` works with CHECK-LABEL, but I am not sure.

qedawkins commented on mlir/test/Dialect/SCF/transform-op-coalesce.mlir:
```diff
@@ -96,3 +101,150 @@ module attributes {transform.with_named_sequence} {
+    %2 = transform.loop.coalesce %1 : (!transform.op<"scf.for">) -> (!transform.op<"scf.for">)
+    transform.yield
+  }
+}
```
ditto: Can you add a CHECK-LABEL here and below? It helps debugging confusing CHECK-NOT related lit test failures (and you can drop the checks for `transform.named_sequence` as well, because I assume that's why those checks are there).

MaheshRavishankar replied:
Actually thats why added `CHECK-NOT` so that there is no match till the `transform.named_sequence`. Ill leave is as `CHECK` for now.

qedawkins reviewed (approved):
LGTM after addressing comments. Might be good to give someone else a chance to check that the new Rewriter API function makes sense, I don't have any context there.

ftynse commented on mlir/lib/Dialect/Affine/Transforms/LoopCoalescing.cpp:
```diff
@@ -35,13 +35,17 @@ namespace {
 struct LoopCoalescingPass
     : public affine::impl::LoopCoalescingBase<LoopCoalescingPass> {
 
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<affine::AffineDialect>();
+  }
```
Why is this necessary? Can this pass produce entities from the affine dialect when none were present in the input?

MaheshRavishankar replied:
Removed now.

ftynse commented on mlir/lib/Dialect/Affine/Utils/LoopUtils.cpp:
```diff
@@ -2765,3 +2765,49 @@ mlir::affine::separateFullTiles(MutableArrayRef<AffineForOp> inputNest,
+  // loops with bounds defined above some loop.
+  // 1. For each loop, find above which parent loop its operands are
+  // defined.
+  SmallVector<unsigned, 4> operandsDefinedAbove(loops.size());
```
Nit: drop the explicit number of stack elements.

ftynse commented on mlir/lib/Dialect/SCF/Transforms/ParallelLoopCollapsing.cpp:
```diff
@@ -8,6 +8,7 @@
 
 #include "mlir/Dialect/SCF/Transforms/Passes.h"
 
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
```
Why is this necessary?

MaheshRavishankar replied:
Removed.

ftynse commented on mlir/lib/Dialect/SCF/Transforms/ParallelLoopCollapsing.cpp:
```diff
@@ -28,6 +29,11 @@ namespace {
 struct TestSCFParallelLoopCollapsing
     : public impl::TestSCFParallelLoopCollapsingBase<
           TestSCFParallelLoopCollapsing> {
+
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<affine::AffineDialect>();
+  }
+
```
We shouldn't have a dependency on affine from SCF.

ftynse commented on mlir/include/mlir/Dialect/SCF/Utils/Utils.h:
```diff
@@ -100,11 +100,16 @@ getSCFMinMaxExpr(Value value, SmallVectorImpl<Value> &dims,
 /// `loops` contains a list of perfectly nested loops with bounds and steps
 /// independent of any loop induction variable involved in the nest.
 LogicalResult coalesceLoops(MutableArrayRef<scf::ForOp> loops);
```
Should we remove the version that doesn't take the rewriter?

MaheshRavishankar replied:
I am fine doing that. Should be a simple fix for downstream users.

MaheshRavishankar replied:
Actually tried it, but this has some strange coupling with things in Affine dialect. I am just going to leave it there for now for people to use at their own risk.

ftynse commented on mlir/lib/Dialect/SCF/Utils/Utils.cpp:
```diff
@@ -12,7 +12,9 @@
 
 #include "mlir/Dialect/SCF/Utils/Utils.h"
 #include "mlir/Analysis/SliceAnalysis.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
```
We shouldn't depend on affine from scf.

ftynse reviewed (changes requested):
SCF must not depend on Affine.

MaheshRavishankar commented:
> SCF must not depend on Affine.

I see your point... but I was trying to use `affine.apply` and `affine.delinearize_index`. Thats why the Affine dialect registration and the dependence come in. Should we just drop that? That would be kind of unfortunate.

MaheshRavishankar commented on mlir/lib/Dialect/Affine/IR/AffineOps.cpp:
```diff
@@ -4532,6 +4532,42 @@ LogicalResult AffineDelinearizeIndexOp::verify() {
 }
 
+namespace {
+// When outer dimension used for delinearization are ones, the corresponding
```
I think so, but wasnt hundred percent sure of the math, but you are right. Ill split this out of this PR though cause I dont need this right now.

MaheshRavishankar commented on mlir/lib/Dialect/SCF/Utils/Utils.cpp:
```diff
@@ -493,62 +520,130 @@ static LoopParams normalizeLoop(OpBuilder &boundsBuilder,
-  inductionVar.replaceAllUsesExcept(shifted, preserve);
-  return {/*lowerBound=*/newLowerBound, /*upperBound=*/newUpperBound,
-          /*step=*/newStep};
+      isStepOne ? diff : rewriter.create<arith::CeilDivSIOp>(loc, diff, step);
```
It should. But doesnt matter now.

MaheshRavishankar commented on mlir/lib/Dialect/SCF/Utils/Utils.cpp:
```diff
@@ -493,62 +520,130 @@ static LoopParams normalizeLoop(OpBuilder &boundsBuilder,
 
-LogicalResult mlir::coalesceLoops(MutableArrayRef<scf::ForOp> loops) {
+/// Helper function to multiply a sequence of values.
+static OpFoldResult getProductOfIndexes(RewriterBase &rewriter, Location loc,
```
Removed uses of affine dialect ops here, so this isnt a problem anymore.

MaheshRavishankar commented on mlir/lib/Dialect/Affine/IR/AffineOps.cpp:
```diff
@@ -4532,6 +4532,42 @@ LogicalResult AffineDelinearizeIndexOp::verify() {
+      return failure();
+    }
+    std::optional<int64_t> basisValue =
+        getConstantIntValue(getAsOpFoldResult(basis.front()));
```
I removed this from this commit. Ill resubmit that later and incorporate this suggestion.

MaheshRavishankar commented on mlir/lib/Dialect/Affine/IR/AffineOps.cpp:
```diff
@@ -4532,6 +4532,42 @@ LogicalResult AffineDelinearizeIndexOp::verify() {
+    replacements.push_back(rewriter.create<arith::ConstantIndexOp>(loc, 0));
+    auto newIndexOp = rewriter.create<AffineDelinearizeIndexOp>(
+        loc, indexOp.getLinearIndex(), basis.drop_front());
+    replacements.append(newIndexOp->result_begin(), newIndexOp->result_end());
```
I removed this from this commit. Ill resubmit that later and incorporate this suggestion.

Commit:
[mlir][SCF] Modernize coalesceLoops method to handle scf.for loops with iter_args

As part of this extension this change also does some general cleanup
1) Make all the methods take RewriterBase as arguments instead of
   creating their own builders that tend to crash when used within
   pattern rewrites
2) Split coalesePerfectlyNestedLoops into two separate methods, one
   for scf.for and other for affine.for. The templatization didnt seem
   to be buying much there.  Also general clean up of tests.

ftynse commented on mlir/lib/Dialect/SCF/Utils/Utils.cpp:
```diff
@@ -557,57 +591,151 @@ LogicalResult mlir::coalesceLoops(MutableArrayRef<scf::ForOp> loops) {
+    if (outerloop.getNumRegionIterArgs() != innerLoop.getNumRegionIterArgs()) {
+      continue;
+    }
+    if (llvm::any_of(
+            llvm::zip_equal(outerloop.getRegionIterArgs(),
+                            innerLoop.getInitArgs()),
+            [](auto it) { return std::get<0>(it) != std::get<1>(it); })) {
+      continue;
+    }
```
Nit: don't we have an `operator==` on `ValueRange` that we could use here?

MaheshRavishankar replied:
Yeah, its unfortunate cause these are BlockList, ValueRange, ArgList and OpResultList... so no direct equality here.

kuhar replied:
we have: `llvm::equal`, it should work here

ftynse replied:
I also suspect that all these are convertible to `ValueRange`, but we may need a explicit conversion constructor or something.

ftynse commented on mlir/test/Dialect/SCF/transform-op-coalesce.mlir:
```diff
@@ -96,3 +102,200 @@ module attributes {transform.with_named_sequence} {
+// CHECK-SAME:     %[[UB2:[a-zA-Z0-9_]+]]: index
+// CHECK-SAME:     %[[STEP2:[a-zA-Z0-9_]+]]: index
+//      CHECK:   %[[NEWUB0_DIFF:.+]] = arith.subi %[[UB0]], %[[LB0]]
+//  CHECK-DAG:   %[[NEWUB0:.+]] = arith.ceildivsi %[[NEWUB0_DIFF]], %[[STEP0]]
+//  CHECK-DAG:   %[[C0:.+]] = arith.constant 0
+//  CHECK-DAG:   %[[C1:.+]] = arith.constant 1
```
Why is this a DAG? Normally, there shouldn't be any non-determinism in the output IR.

joker-eph replied:
DAG isn't about non-determinism (otherwise it wouldn't exist at all, since as you mention there shouldn't be any). Instead it can indicate portion where "we don't care about the order". In particular for constant, we have been changing the strategy of "CSE on the fly" quite a few times, and it's annoying when test encode an order while it isn't relevant to what is actually tested.

ftynse reviewed (approved):
Thanks for iterating! LGTM modulo some nits.

MaheshRavishankar commented on mlir/lib/Dialect/SCF/Utils/Utils.cpp:
```diff
@@ -472,18 +473,23 @@ LogicalResult mlir::loopUnrollByFactor(
+/// expected to be either `loop` or another loop perfectly nested under `loop`.
+/// Insert the definition of new bounds immediate before `outer`, which is
+/// expected to be either `loop` or its parent in the loop nest.
+static LoopParams normalizeLoop(RewriterBase &rewriter, Location loc, Value lb,
```
Actually it is doing it in place. So maybe normalizeLoops is better.

ftynse replied:
Hmm, I don't see _this_ function doing it. The caller (`coalesceLoops`) does the in-place around line 600.

MaheshRavishankar commented on mlir/lib/Dialect/SCF/Utils/Utils.cpp:
```diff
@@ -557,57 +591,151 @@ LogicalResult mlir::coalesceLoops(MutableArrayRef<scf::ForOp> loops) {
+    Operation *innerTerminator = innerLoop.getBody()->getTerminator();
+    auto yieldedVals = llvm::to_vector(innerTerminator->getOperands());
+    rewriter.eraseOp(innerTerminator);
+    outerLoop.getBody()->getOperations().splice(
+        Block::iterator(innerLoop), innerLoop.getBody()->getOperations());
```
Good catch. Thanks!
